File: scripts/build_product_registry.js
const fs = require("fs");
const path = require("path");

const WEB_DIR = path.resolve(__dirname, "..");
const ROOT = path.resolve(WEB_DIR, "..", "..");
const REGISTRY_DIR = path.join(ROOT, "_registry");
const PRODUCTS_DIR = path.join(REGISTRY_DIR, "products");
const OUT = path.join(REGISTRY_DIR, "product_registry.json");

function readJson(file) {
  return JSON.parse(fs.readFileSync(file, "utf8"));
}

function writeJson(file, value) {
  fs.writeFileSync(file, `${JSON.stringify(value, null, 2)}\n`, "utf8");
}

function rel(file) {
  return path.relative(ROOT, file).replace(/\\/g, "/");
}

function listJsonFiles(dir) {
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir, { withFileTypes: true })
    .filter((entry) => entry.isFile() && /\.products\.json$/i.test(entry.name))
    .map((entry) => path.join(dir, entry.name))
    .sort((a, b) => a.localeCompare(b));
}

function asArray(value) {
  return Array.isArray(value) ? value : [];
}

function assertObject(value, message) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(message);
  }
}

function loadCompanyIds() {
  const file = path.join(REGISTRY_DIR, "company_registry.json");
  const registry = readJson(file);
  return new Set(Object.keys(registry.companies || {}));
}

function loadStudioIds() {
  const file = path.join(REGISTRY_DIR, "studio_registry.json");
  const registry = fs.existsSync(file) ? readJson(file) : { studios: {} };
  return new Set(Object.keys(registry.studios || {}));
}

function validateProduct(productId, product, sourceFile, companyIds, studioIds) {
  const errors = [];
  const required = [
    "product_id",
    "name_cn",
    "name_en",
    "aliases",
    "developer_company_ids",
    "publisher_company_ids",
    "studio_ids",
    "obsidian_link",
  ];

  for (const key of required) {
    if (!(key in product)) errors.push(`${productId}: missing ${key}`);
  }
  if (product.product_id !== productId) {
    errors.push(`${productId}: product_id field is ${JSON.stringify(product.product_id)}`);
  }

  for (const key of ["aliases", "developer_company_ids", "publisher_company_ids", "studio_ids"]) {
    if (!Array.isArray(product[key])) errors.push(`${productId}: ${key} must be an array`);
  }

  for (const companyId of [...asArray(product.developer_company_ids), ...asArray(product.publisher_company_ids)]) {
    if (!companyIds.has(companyId)) errors.push(`${productId}: unknown company_id ${companyId}`);
  }

  for (const studioId of asArray(product.studio_ids)) {
    if (!studioIds.has(studioId)) errors.push(`${productId}: unknown studio_id ${studioId}`);
  }

  if (errors.length) {
    const prefix = rel(sourceFile);
    throw new Error(errors.map((error) => `${prefix}: ${error}`).join("\n"));
  }
}

function buildRegistry() {
  const companyIds = loadCompanyIds();
  const studioIds = loadStudioIds();
  const files = listJsonFiles(PRODUCTS_DIR);
  const products = {};
  const sourceFiles = [];

  for (const file of files) {
    const source = readJson(file);
    assertObject(source.products, `${rel(file)}: products must be an object`);
    const sourceCompanyId = source.source_company_id || path.basename(file, ".products.json");
    if (!companyIds.has(sourceCompanyId)) {
      throw new Error(`${rel(file)}: unknown source_company_id ${sourceCompanyId}`);
    }
    sourceFiles.push(rel(file));

    for (const [productId, product] of Object.entries(source.products)) {
      if (products[productId]) {
        throw new Error(`${rel(file)}: duplicate product_id ${productId}`);
      }
      validateProduct(productId, product, file, companyIds, studioIds);
      products[productId] = {
        product_id: productId,
        source_company_id: sourceCompanyId,
        ...product,
        product_id: productId,
      };
    }
  }

  const registry = {
    schema_version: "1.0",
    updated_at: new Date().toISOString().slice(0, 10),
    source_files: sourceFiles,
    products,
  };
  writeJson(OUT, registry);
  console.log(`Wrote ${rel(OUT)} (${Object.keys(products).length} products from ${files.length} files)`);
}

buildRegistry();

File: scripts/build_product_registry.py
import json
from datetime import date
from pathlib import Path


WEB_DIR = Path(__file__).resolve().parents[1]
ROOT = WEB_DIR.parents[1]
REGISTRY_DIR = ROOT / "_registry"
PRODUCTS_DIR = REGISTRY_DIR / "products"
OUT = REGISTRY_DIR / "product_registry.json"


def read_json(path):
    return json.loads(path.read_text(encoding="utf-8"))


def write_json(path, value):
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def rel(path):
    return path.relative_to(ROOT).as_posix()


def as_list(value):
    return value if isinstance(value, list) else []


def load_company_ids():
    registry = read_json(REGISTRY_DIR / "company_registry.json")
    return set(registry.get("companies", {}).keys())


def load_studio_ids():
    path = REGISTRY_DIR / "studio_registry.json"
    if not path.exists():
        return set()
    registry = read_json(path)
    return set(registry.get("studios", {}).keys())


def validate_product(product_id, product, source_file, company_ids, studio_ids):
    required = [
        "product_id",
        "name_cn",
        "name_en",
        "aliases",
        "developer_company_ids",
        "publisher_company_ids",
        "studio_ids",
        "obsidian_link",
    ]
    errors = []

    for key in required:
        if key not in product:
            errors.append(f"{product_id}: missing {key}")
    if product.get("product_id") != product_id:
        errors.append(f"{product_id}: product_id field is {product.get('product_id')!r}")

    for key in ["aliases", "developer_company_ids", "publisher_company_ids", "studio_ids"]:
        if key in product and not isinstance(product[key], list):
            errors.append(f"{product_id}: {key} must be an array")

    for company_id in as_list(product.get("developer_company_ids")) + as_list(product.get("publisher_company_ids")):
        if company_id not in company_ids:
            errors.append(f"{product_id}: unknown company_id {company_id}")

    for studio_id in as_list(product.get("studio_ids")):
        if studio_id not in studio_ids:
            errors.append(f"{product_id}: unknown studio_id {studio_id}")

    if errors:
        prefix = rel(source_file)
        raise ValueError("\n".join(f"{prefix}: {error}" for error in errors))


def build_registry():
    company_ids = load_company_ids()
    studio_ids = load_studio_ids()
    products = {}
    source_files = []

    for source_file in sorted(PRODUCTS_DIR.glob("*.products.json")):
        source = read_json(source_file)
        source_products = source.get("products")
        if not isinstance(source_products, dict):
            raise ValueError(f"{rel(source_file)}: products must be an object")

        source_company_id = source.get("source_company_id") or source_file.name.removesuffix(".products.json")
        if source_company_id not in company_ids:
            raise ValueError(f"{rel(source_file)}: unknown source_company_id {source_company_id}")

        source_files.append(rel(source_file))
        for product_id, product in source_products.items():
            if product_id in products:
                raise ValueError(f"{rel(source_file)}: duplicate product_id {product_id}")
            validate_product(product_id, product, source_file, company_ids, studio_ids)
            products[product_id] = {
                "product_id": product_id,
                "source_company_id": source_company_id,
                **product,
                "product_id": product_id,
            }

    registry = {
        "schema_version": "1.0",
        "updated_at": date.today().isoformat(),
        "source_files": source_files,
        "products": products,
    }
    write_json(OUT, registry)
    print(f"Wrote {rel(OUT)} ({len(products)} products from {len(source_files)} files)")


if __name__ == "__main__":
    build_registry()
